src/contracts/middleware/BLSRegistryCoordinatorWithIndices.sol

```diff
@@ -0,0 +1,555 @@
+            updateBlockNumber: uint32(block.number),
+            nextUpdateBlockNumber: 0,
+            quorumBitmap: uint192(quorumBitmap)
+        }));
+
+        // set the operator struct
+        _operators[operator] = Operator({
+            operatorId: operatorId,
+            status: OperatorStatus.REGISTERED
+        });
+
+        // record a stake update not bonding the operator at all (unbonded at 0), because they haven't served anything yet
+        // serviceManager.recordFirstStakeUpdate(operator, 0);
+
+        emit OperatorRegistered(operator, operatorId);
+
+        emit OperatorSocketUpdate(operatorId, socket);
+
+        return numOperatorsPerQuorum;
+    }
+
+    function _registerOperatorWithCoordinatorAndNoOverfilledQuorums(address operator, bytes calldata quorumNumbers, BN254.G1Point memory pubkey, string memory socket) internal {
+        uint32[] memory numOperatorsPerQuorum = _registerOperatorWithCoordinator(operator, quorumNumbers, pubkey, socket);
+        for (uint i = 0; i < numOperatorsPerQuorum.length; i++) {
+            require(
+                numOperatorsPerQuorum[i] <= _quorumOperatorSetParams[uint8(quorumNumbers[i])].maxOperatorCount,
+                "BLSRegistryCoordinatorWithIndices._registerOperatorWithCoordinatorAndNoOverfilledQuorums: quorum is overfilled"
+            );
+        }
+    }
+
+    function _deregisterOperatorWithCoordinator(address operator, bytes calldata quorumNumbers, BN254.G1Point memory pubkey, bytes32[] memory operatorIdsToSwap) internal {
+        require(_operators[operator].status == OperatorStatus.REGISTERED, "BLSRegistryCoordinatorWithIndices._deregisterOperatorWithCoordinator: operator is not registered");
+
+        // get the operatorId of the operator
+        bytes32 operatorId = _operators[operator].operatorId;
+        require(operatorId == pubkey.hashG1Point(), "BLSRegistryCoordinatorWithIndices._deregisterOperatorWithCoordinator: operatorId does not match pubkey hash");
+
+        // get the quorumNumbers of the operator
+        uint256 quorumsToRemoveBitmap = BitmapUtils.orderedBytesArrayToBitmap(quorumNumbers);
+        require(quorumsToRemoveBitmap <= MiddlewareUtils.MAX_QUORUM_BITMAP, "BLSRegistryCoordinatorWithIndices._deregisterOperatorWithCoordinator: quorumsToRemoveBitmap exceeds of max bitmap size");
+        uint256 operatorQuorumBitmapHistoryLengthMinusOne = _operatorIdToQuorumBitmapHistory[operatorId].length - 1;
+        uint192 quorumBitmapBeforeUpdate = _operatorIdToQuorumBitmapHistory[operatorId][operatorQuorumBitmapHistoryLengthMinusOne].quorumBitmap;
+        // check that the quorumNumbers of the operator matches the quorumNumbers passed in
+        require(
+            quorumBitmapBeforeUpdate & quorumsToRemoveBitmap == quorumsToRemoveBitmap,
+            "BLSRegistryCoordinatorWithIndices._deregisterOperatorWithCoordinator: cannot deregister operator for quorums that it is not a part of"
+        );
```

Comment:
going to point out a potential race condition here -- if an operator deregisters from a single quorum, they can potentially block themselves from being "ejected" from other quorums (or at least cause a call to eject them to revert). doesn't seem like a major issue, but there's definitely a number of similar race conditions here.

Reply:
seems like this particular race condition could be fixed by setting `quorumsToRemoveBitmap = quorumBitmapBeforeUpdate & quorumsToRemoveBitmap` (i.e. filtering out quorums that the ejected operator isn't in), rather than the strict require statement here.
Going to recommend that change, since I think the race condition is ~low-medium bad as-is.
@0x0aa0 friendly ping on this

Reply:
I think my suggestion might be broken, since we pass on the input from which `quorumsToRemoveBitmap` is derived -- might need to do something like
```
if( quorumBitmapBeforeUpdate & quorumsToRemoveBitmap != quorumsToRemoveBitmap ) {
 use the BitmapUtils library to take the quorumBitmapBeforeUpdate and derive the appropriate 'bytes' object, then pass that on.
}
```